### PR TITLE
An optional id in the subscribe request to match a subscription with newly created elements in the stream list

### PR DESCRIPTION
**Pull request by JanFellner**
Followup for this one: https://github.com/meetecho/janus-gateway/issues/3024
Comments welcome
(changed from private janus fork to estos fork to better follow the changes on our side)

### CODE REVIEW

File: src/plugins/janus_videoroom.c [hunk 1]
```diff
@@ -2205,6 +2206,7 @@ typedef struct janus_videoroom_subscriber_stream {
 	GSList *publisher_streams;						/* Complete list of publisher streams (e.g., when this is data) */
 	int mindex;				/* The media index of this stream (may not be the same as the publisher stream) */
 	char *mid;				/* The mid of this stream (may not be the same as the publisher stream) */
+	char* opaqueid;			/* An id provided while subscribing to uniquely identify the subscription in the updated events */
```
Comment (lminiero) [inline src/plugins/janus_videoroom.c:2209]: Code style: * next to the pointer, not the type.

File: src/plugins/janus_videoroom.c [hunk 2]
```diff
@@ -2991,6 +2994,7 @@ static void janus_videoroom_create_dummy_publisher(janus_videoroom *room, GHashT
 /* Helpers for subscription streams */
 static janus_videoroom_subscriber_stream *janus_videoroom_subscriber_stream_add(janus_videoroom_subscriber *subscriber,
 		janus_videoroom_publisher_stream *ps,
+		const char* opaqueid,
```
Comment (lminiero) [inline src/plugins/janus_videoroom.c:2997]: See above. Besides, per code style, this should be `opaque_id', not `opaqueid`.

File: src/plugins/janus_videoroom.c [hunk 3]
```diff
@@ -3003,6 +3007,7 @@ static janus_videoroom_subscriber_stream *janus_videoroom_subscriber_stream_add(
 	/* Allocate a new subscriber stream instance */
 	janus_videoroom_subscriber_stream *stream = g_malloc0(sizeof(janus_videoroom_subscriber_stream));
 	stream->mindex = g_list_length(subscriber->streams);
+	stream->opaqueid = g_strdup(opaqueid);
```
Comment (lminiero) [inline src/plugins/janus_videoroom.c:3010]: This assumes `opaqueid` is always non-NULL, which is wrong. You should only duplicate the string if it exists.

File: src/plugins/janus_videoroom.c [hunk 4]
```diff
@@ -3051,7 +3056,8 @@ static janus_videoroom_subscriber_stream *janus_videoroom_subscriber_stream_add(
 }
 
 static janus_videoroom_subscriber_stream *janus_videoroom_subscriber_stream_add_or_replace(janus_videoroom_subscriber *subscriber,
-		janus_videoroom_publisher_stream *ps) {
+		janus_videoroom_publisher_stream *ps,
+		const char* opaqueid) {
```
Comment (lminiero) [inline src/plugins/janus_videoroom.c:3060]: See above,

File: src/plugins/janus_videoroom.c [hunk 5]
```diff
@@ -3111,10 +3117,12 @@ static janus_videoroom_subscriber_stream *janus_videoroom_subscriber_stream_add_
 		}
 		temp = temp->next;
 	}
-	if(found)
+	if(found)  {
+		stream->opaqueid = g_strdup(opaqueid);
```
Comment (lminiero) [inline src/plugins/janus_videoroom.c:3121]: See above. Besides, you're introducing leaks if `stream->opaqueid` exists already, as you're replacing the pointer and the old one will never be freed.

File: src/plugins/janus_videoroom.c [hunk 6]
```diff
@@ -8458,6 +8468,7 @@ static void *janus_videoroom_handler(void *data) {
 					}
 					janus_mutex_lock(&publisher->streams_mutex);
 					const char *mid = json_string_value(json_object_get(s, "mid"));
+					const char *opaqueid = json_string_value(json_object_get(s, "opaque_id"));
```
Comment (lminiero) [inline src/plugins/janus_videoroom.c:8471]: New attributes that are added must be added to the related `_parameters` validation too.

File: src/plugins/janus_videoroom.c [hunk 7]
```diff
@@ -9459,6 +9470,7 @@ static void *janus_videoroom_handler(void *data) {
 						}
 						/* Are we subscribing to this publisher as a whole or only to a single stream? */
 						const char *mid = json_string_value(json_object_get(s, "mid"));
+						const char *opaqueid = json_string_value(json_object_get(s, "opaque_id"));
```
Comment (lminiero) [inline src/plugins/janus_videoroom.c:9473]: Same here.

File: src/plugins/janus_videoroom.c [hunk 8]
```diff
@@ -3168,6 +3176,8 @@ static json_t *janus_videoroom_subscriber_streams_summary(janus_videoroom_subscr
 		json_object_set_new(m, "active", (ps || stream->type == JANUS_VIDEOROOM_MEDIA_DATA) ? json_true() : json_false());
 		json_object_set_new(m, "mindex", json_integer(stream->mindex));
 		json_object_set_new(m, "mid", json_string(stream->mid));
+		if(stream->opaqueid)
+			json_object_set_new(m, "opaque_id", json_string(stream->opaqueid));
```
Comment (lminiero) [inline src/plugins/janus_videoroom.c:3180]: I'm not sure this should always be here. `janus_videoroom_subscriber_streams_summary` is used for different targets, not just notifying updates back to the subscriber, and this opaque identifier feels like something that only belongs to that subscriber's events.

**Review by lminiero — COMMENTED** [review]
I had a quick glance. First of all, I'm not in love with the `opaque_id` terminology, mostly because it exists already in the core and will cause confusion when used in a plugin as well. Maybe it can just be something as simple as `description`, which is similar to what we have for publisher streams: since a `description` can be anything, it can be used for display purposes if you want, or for correlation purposes if you want. This is especially relevant considering you call it an ID, but looking at the code it can change any time you update a subscription, meaning it's probably not an ID but more something like metadata.

That said, I understand the point, but I'm still not 100% sure it would help you solve the problem, especially if this opaque property is changed across requests for the same m-line. Since as we discussed you may not receive updates in response to each subscribe/unsubscribe/update you send, but may receive less events that "group" the effect of multiple requests: if different opaque values were passed for the same m-line across different requests in sequence, you'll only get the last one in the event, so you may be waiting on an opaque you'll never receive.

I added a few additional notes on things that need to be fixed in the code too.

**Comment by JanFellner** [comment]
Hi Lorenzo,

thanks for the review, before stepping into the code let´s shortly discuss the general topics above the code :).

> I had a quick glance. First of all, I'm not in love with the `opaque_id` terminology, mostly because it exists already in the core and will cause confusion when used in a plugin as well. Maybe it can just be something as simple as `description`, which is similar to what we have for publisher streams: since a `description` can be anything, it can be used for display purposes if you want, or for correlation purposes if you want. This is especially relevant considering you call it an ID, but looking at the code it can change any time you update a subscription, meaning it's probably not an ID but more something like metadata.

I followed the opaque_id as it did express exactly the wanted purpose to me (https://github.com/meetecho/janus-gateway/pull/748). I am fine with any name. Reusing the same is fine to me. I would otherwise call it cross ref id (as also used in csta context when monitoring something).
Just tell me if we should reuse description which is already existing (max also cause confusion as the description is then two times in the event, feed_description and description) or if we should have an own name for that purpose. The suggestion from my side would then be **crossrefid**.
If it is currently possible to change that attribute in an update request this is a non wanted behaviour. I see it really as correlation id between the subscription request and the first event that hands back the ids on the subscriber peerconnection but also in future events to be able to simplier align the janus event payload with own lookup maps.

> That said, I understand the point, but I'm still not 100% sure it would help you solve the problem, especially if this opaque property is changed across requests for the same m-line. Since as we discussed you may not receive updates in response to each subscribe/unsubscribe/update you send, but may receive less events that "group" the effect of multiple requests: if different opaque values were passed for the same m-line across different requests in sequence, you'll only get the last one in the event, so you may be waiting on an opaque you'll never receive.
- It should not be possible to change it. If thats currently the case i need to check where on an update this is overwritten.
- No receiving upates to all unique operations i call is fine to me, the logic does not care.
- The property 100% matches the need to be able to identify the subscribe operation with following events, no matter if with the next or any consequent following.

> I added a few additional notes on things that need to be fixed in the code too.
I would go over the code once we have the basic constraints fixed (naming etc.)

Just to have it mentioned:
g_strdup returns null if the argument is null. You want to have that checked with an if (argument) g_strup i can do that, technically it´s not needed but ensures that the variable it points to is properly nulled.

File: src/plugins/janus_videoroom.c (hunk 1 above)
Comment (JanFellner) [reply src/plugins/janus_videoroom.c]: fixed

File: src/plugins/janus_videoroom.c (hunk 8 above)
Comment (JanFellner) [reply src/plugins/janus_videoroom.c]: I see the place actually exactly here. This is the element in the streams list i cannot associated if i don´t have the cross reference id with the subscribe event. Thus this is the idea of the PR to have the correlation id in the streams list as provided in the subscribe request.

File: src/plugins/janus_videoroom.c (hunk 6 above)
Comment (JanFellner) [reply src/plugins/janus_videoroom.c]: Am i missing something? It is added in the subscriber_stream_parameters object...

File: src/plugins/janus_videoroom.c (hunk 7 above)
Comment (JanFellner) [reply src/plugins/janus_videoroom.c]: Same here.
Am i missing something? It is added in the subscriber_stream_parameters object...

**Comment by JanFellner** [comment]
Changed to ready for review, running now "produtive" on our dev environment

File: src/plugins/janus_videoroom.c [hunk 9]
```diff
@@ -2991,6 +2996,7 @@ static void janus_videoroom_create_dummy_publisher(janus_videoroom *room, GHashT
 /* Helpers for subscription streams */
 static janus_videoroom_subscriber_stream *janus_videoroom_subscriber_stream_add(janus_videoroom_subscriber *subscriber,
 		janus_videoroom_publisher_stream *ps,
+		const char *crossrefid,
```
Comment (lminiero) [inline src/plugins/janus_videoroom.c:2999]: No need for a new line, just move after ps.

**Review by lminiero — COMMENTED** [review]
Thanks for the update. I'll be busy at the IETF meeting until the end of the month, so I don't think I'll have time to test this before I come back: I added just a couple of notes from a quick glance at the code. I'll let @atoppi do a deeper review.

File: src/plugins/janus_videoroom.c [hunk 10]
```diff
@@ -3051,7 +3058,8 @@ static janus_videoroom_subscriber_stream *janus_videoroom_subscriber_stream_add(
 }
 
 static janus_videoroom_subscriber_stream *janus_videoroom_subscriber_stream_add_or_replace(janus_videoroom_subscriber *subscriber,
-		janus_videoroom_publisher_stream *ps) {
+		janus_videoroom_publisher_stream *ps,
+		const char *crossrefid) {
```
[the remark this answers is not in the document]
Comment (JanFellner) [reply src/plugins/janus_videoroom.c]: fixed

File: src/plugins/janus_videoroom.c [hunk 11]
```diff
@@ -3111,10 +3119,14 @@ static janus_videoroom_subscriber_stream *janus_videoroom_subscriber_stream_add_
 		}
 		temp = temp->next;
 	}
-	if(found)
+	if(found)  {
+		if(stream->crossrefid)
+			g_free(stream->crossrefid);
```
[the remark this answers is not in the document]
Comment (JanFellner) [reply src/plugins/janus_videoroom.c]: fixed

File: src/plugins/janus_videoroom.c (hunk 9 above)
Comment (JanFellner) [reply src/plugins/janus_videoroom.c]: fixed

**Comment by atoppi** [comment]
Changes look good to me.
Tested with a modified version of `mvideoroomtest.js` and it seems to work as expected with the `update` API.

@JanFellner According to your use cases, I'm wondering if it would make any sense to add this `crossrefid` to the `streams` list when joining a new subscriber.

```
{
  "janus": "message",
  "body": {
    "request": "join",
    "room": 1234,
    "ptype": "subscriber",
    "streams": [
      {
        "feed": 4446040227822893,
        "mid": "0"
      },
      {
        "feed": 4446040227822893,
        "mid": "1"
      }
    ],
    "private_id": 2847518190
  },
  "transaction": "AvkXpeLdGoLZ",
  "session_id": 431131763148161,
  "handle_id": 7854428491773330
}
```

**Comment by JanFellner** [comment]
@atoppi maybe i am missing the core question but the crossrefid is added from the client calling janus in exactly that request and returend in any stream related event from janus.

This is how such a request and response looks like.

<pre>
{
  "_": "2022-07-25T11:00:17.345Z - ws - out",
  "janus": "message",
  "transaction": "S_message_7_6",
  "handle_id": 6602915452921274,
  "session_id": 7047460590128940,
  "timeout": 500,
  "body": {
    "request": "join",
    "ptype": "subscriber",
    "room": "R_55698ee42ed94463ae6b0ae9ace87fa7",
    "streams": [
      {
        "feed": "56174d8c-d6e2-48ae-ad6b-7da87e973362",
        "mid": "0",
        "crossrefid": "S_82575842511f4d39ac2b88b5722b0d79"
      },
      {
        "feed": "56174d8c-d6e2-48ae-ad6b-7da87e973362",
        "mid": "1",
        "crossrefid": "S_82575842511f4d39ac2b88b5722b0d79",
        "substream": 2,
        "temporal": 1
      }
    ]
  },
  "opaque_id": "P_bea32134fe0c44c3bee95cfced8c133c|publisher"
}
</pre>

<pre>
{
  "_": "2022-07-25T11:00:17.388Z - ws - in",
  "janus": "event",
  "session_id": 7047460590128940,
  "transaction": "S_message_7_6",
  "sender": 6602915452921274,
  "plugindata": {
    "plugin": "janus.plugin.videoroom",
    "data": {
      "videoroom": "attached",
      "room": "R_55698ee42ed94463ae6b0ae9ace87fa7",
      "streams": [
        {
          "type": "audio",
          "active": true,
          "mindex": 0,
          "mid": "0",
          "crossrefid": "S_82575842511f4d39ac2b88b5722b0d79",
          "ready": false,
          "send": true,
          "feed_id": "56174d8c-d6e2-48ae-ad6b-7da87e973362",
          "feed_display": "P_8f248c81a2174b10a2153eed2a483152",
          "feed_mid": "0",
          "feed_description": "Video 2 - left human main",
          "codec": "opus"
        },
        {
          "type": "video",
          "active": true,
          "mindex": 1,
          "mid": "1",
          "crossrefid": "S_82575842511f4d39ac2b88b5722b0d79",
          "ready": false,
          "send": true,
          "feed_id": "56174d8c-d6e2-48ae-ad6b-7da87e973362",
          "feed_display": "P_8f248c81a2174b10a2153eed2a483152",
          "feed_mid": "1",
          "feed_description": "Video 2 - left human main",
          "codec": "vp8",
          "simulcast": {
            "substream": -1,
            "substream-target": 2,
            "temporal-layer": -1,
            "temporal-layer-target": 1
          }
        }
      ]
    }
  },
  "jsep": {
    "type": "offer",
    "sdp": "..."
  }
}
</pre>

**Comment by lminiero** [comment]
Yes, it's a per-subscriber property, not per-publisher from what I understood.

**Comment by JanFellner** [comment]
In my understanding / terminology a per subscription id. 
Cause i als a subscriber join one time but hold 1:n subscriptions on different feeds in the room.
To differ these feeds, especially in case where i hold multiple subscriptions on the same feed, i need this id.

**Comment by atoppi** [comment]
> @atoppi maybe i am missing the core question but the crossrefid is added from the client calling janus in exactly that request and returend in any stream related event from janus.

You're right, I forgot to customize the JS client in one point, sorry for the noise.
Anyway the documentation [is still missing ](https://github.com/ESTOS/janus-gateway/blob/feature/3024-add-opaque-id-to-subscriptions/src/plugins/janus_videoroom.c#L1051)the `crossrefid` parameter in the `join` as a subscriber section.

**Comment by JanFellner** [comment]
Fixed :)
https://github.com/ESTOS/janus-gateway/blob/feature/3024-add-opaque-id-to-subscriptions/src/plugins/janus_videoroom.c#L1055

**Comment by JanFellner** [comment]
So *whom* clicks merge now 😆

**Comment by lminiero** [comment]
Ack, thanks for the patch and for your patience, merging now :+1:

**Pull request merged**